db2i_charsetSupport.cc

```diff
@@ -759,10 +423,76 @@ int32 getConversion(enum_conversionDirection direction, const CHARSET_INFO* cs,
                 sizeof(hashKey))))
   {
     DBUG_PRINT("getConversion", ("Hash miss for direction=%d, cs=%s, ccsid=%d", direction, cs->name, db2CCSID));
-    rc= openNewConversion(direction, cs->csname, db2CCSID, &hashKey, conversion);
-    if (rc)
-      DBUG_RETURN(rc);
-  }
+
+    EncodingMapEntry *entry = nullptr;
+
+    for (auto& e : encoding_map) {
+      if (strcmp(e.csname, cs->csname) == 0) {
+        entry = &e;
+        break;
+      }
+    }
+
+    if (entry == nullptr) {
+      fprintf(stderr, "ibmdb2 error entry was null charset was not found\n");
+      getErrTxt(DB2I_ERR_UNSUPP_CHARSET, cs->csname);
+      DBUG_RETURN(DB2I_ERR_UNSUPP_CHARSET);
+    }
+
+    const char *iconv_db2 = entry->iconv_db2;
+    // IBM-12345 is the max length of temp buffer
+    char temp[10];
+    if (db2CCSID != entry->db2_ccsid) {
+      switch(db2CCSID) {
+        case 1208:
+        iconv_db2 = "UTF-8";
+        break;
+        case 1200:
+        iconv_db2 = "UTF-16";
+        break;
+        case 13488:
+        iconv_db2 = "UCS-2";
+        break;
+        default:
+        sprintf(temp, "IBM-%hu", db2CCSID);
+        iconv_db2 = temp;
+      }
+    }
+
+    iconv_t newConversion;
+    /* Call iconv to open the conversion. */
+    if (direction == toDB2)
+    {
+      newConversion = iconv_open(iconv_db2, entry->iconv_mariadb);
+    }
+    else
+    {
+      newConversion = iconv_open(entry->iconv_mariadb, iconv_db2);
```

Comment:
I don't see an `iconv_close()` anywhere (pre-existing problem?)

Reply:
Here it opens converters and caches it. Looks like iconv_close is handled in [dmap_open](https://github.com/zendtech/db2iengine/blob/2384b2587f31ff925ddba9d6ed8f3737866331b4/db2i_myconv.cc#L151)

Reply:
@ThePrez FYI, iconv_open/iconv/iconv_close has been re-defined to myconv_open/myconv_dmap/myconv_close here: https://github.com/zendtech/db2iengine/blob/master/db2i_iconv.h#L47-L49. The _actual_ iconv handles are only used in dmap_open, where they are opened, used to generate the internal dmap conversion table, then closed.